builder/bundles.go

```diff
@@ -480,33 +515,141 @@ func downloadRpms(packagerCmd, rpmList []string, baseDir string, maxRetries int)
+	rpm2Cmd.Dir = dir
+	rpm2Cmd.Env = os.Environ()
+
+	rpmTar := file + ".tgz"
+	tarCmd := exec.Command("tar", "-xf", rpmTar, "-C", baseDir)
+	tarCmd.Env = os.Environ()
+	tarCmd.Dir = dir
+
+	var err error
+
+	err = rpm2Cmd.Run()
+	// ignoring 255 error code for now, till next release of clear which fixes the bug of rpm2archive
+	if err != nil && err.Error() != "exit status 255" {
+		return fmt.Errorf("rpm2archive cmd failed with %v", err.Error())
+	}
+
+	err = tarCmd.Run()
+	if err != nil {
+		return fmt.Errorf("tarCmd failed with %v", err)
+	}
+
+	err = os.Remove(dir + rpmTar)
+	if err != nil {
+		fmt.Println("failed to remove file")
+	}
+
+	return nil
+}
+
+func installBundleToFull(packagerCmd []string, baseDir string, bundle *bundle, downloadRetries int, numWorkers int, localPath string) error {
 	var err error
-	baseDir := filepath.Join(buildVersionDir, "full")
-	rpmList := []string{}
+	var wg sync.WaitGroup
+	rpmCh := make(chan string)
+
+	var errCh error
+	var rpmPath string
+	var missingRpms []string
 
-	if len(bundle.AllPackages) > 0 {
-		// There were packages directly included for this bundle so
-		// install to full chroot. This check is necessary so we don't
-		// call dnf install with no package listed.
-		for p := range bundle.AllPackages {
-			rpmList = append(rpmList, p)
+	if len(bundle.AllRpmPackages) < numWorkers {
+		numWorkers = len(bundle.AllRpmPackages)
+	}
+	errorCh := make(chan error, numWorkers)
+	wg.Add(numWorkers)
+
+	rpmWorker := func() {
+		var e error
+		for rpm := range rpmCh {
+			// running for extractRetries times as rpm2archive and tar can failed if two files are extracted at exact same path
+			for i := 0; i < extractRetries; i++ {
+				e = extractRpm(baseDir, rpm)
+				if e == nil {
+					break
+				}
+			}
+			if e != nil {
+				errorCh <- e
+				break
+			}
 		}
+		wg.Done()
+	}
 
-		// Retry RPM downloads to avoid timeout failures due to slow network
-		_, err = downloadRpms(packagerCmd, rpmList, baseDir, downloadRetries)
-		if err != nil {
-			return err
+	for rpm := range bundle.AllRpmPackages {
+		rpmFull := filepath.Join(localPath, rpm)
+		if _, err = os.Stat(rpmFull); os.IsNotExist(err) {
+			rpmName := strings.TrimSuffix(rpm, ".rpm")
+			missingRpms = append(missingRpms, rpmName)
 		}
+	}
 
-		args := merge(packagerCmd, "--installroot="+baseDir, "install")
-		args = append(args, rpmList...)
-		_, err = helpers.RunCommandOutputEnv(args[0], args[1:], []string{"LC_ALL=en_US.UTF-8"})
-		if err != nil {
-			return err
+	if len(missingRpms) > 0 {
+		// downloadRpm if not in offline mode, if in offline mode return error
+		if !Offline {
+			rpmPath, err = downloadRpm(packagerCmd, missingRpms, baseDir, downloadRetries)
+			if err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("rpms not found: %s ", missingRpms)
+		}
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go rpmWorker()
+	}
+
+	// feed the channel
+	for rpm := range bundle.AllRpmPackages {
+		if rpmMap[rpm] {
```

Comment:
Is there ever a situation where this will be true other than for `filesystem`? From what I can see, since `rpm` is the map key, this loop will never loop twice on the same value, so other than for `filesystem` that is processed beforehand, this condition is never reached.

Reply:
Ideally not but it is possible to have subpackages where the content is in a folder of another subpackage. Would be something we'd want scanned prior to release generation I think.

Though actually we have another option, after the extraction we could do permissions checks/fixups based on what the rpm says the files it contains should have. That would let us skip the filesystem must be first hack too.

Reply:
That would make the process slower though so it isn't ideal.

Reply:
I'm not sure I follow, but I don't think this applies here. We are iterating on rpm package names, their content should not interfere at all on this loop.

Reply:
Oh nevermind, I realize my confusion now. This is indeed required.